Index: tools/clang/include/clang/Basic/DiagnosticSemaKinds.td
===================================================================
--- tools/clang/include/clang/Basic/DiagnosticSemaKinds.td	(revision 188057)
+++ tools/clang/include/clang/Basic/DiagnosticSemaKinds.td	(working copy)
@@ -30,6 +30,13 @@
   "been assigned">, InGroup<DiagGroup<"duplicate-enum">>, DefaultIgnore;
 def note_duplicate_element : Note<"element %0 also has value %1">;
 
+// Automatic Differentiation
+def err_differentiating_undefined_function : Error<
+  "attempted differention of function '%0', which does not have a definition">;
+def warn_function_not_declared_in_custom_derivatives : Warning<
+  "function '%0' was not differentiated because it is not declared in "
+  "namespace 'custom_derivatives'">;
+
 // Constant expressions
 def err_expr_not_ice : Error<
   "expression is not an %select{integer|integral}0 constant expression">;
