Index: tools/CMakeLists.txt
===================================================================
--- tools/CMakeLists.txt	(revision 182931)
+++ tools/CMakeLists.txt	(working copy)
@@ -58,6 +58,7 @@
   endif()
 endif()
 
+add_llvm_external_project(autodiff)
 add_llvm_external_project(clang)
 add_llvm_external_project(lld)
 add_llvm_external_project(lldb)
Index: tools/Makefile
===================================================================
--- tools/Makefile	(revision 182931)
+++ tools/Makefile	(working copy)
@@ -21,7 +21,7 @@
 
 # Build LLDB if present. Note LLDB must be built last as it depends on the
 # wider LLVM infrastructure (including Clang).
-OPTIONAL_DIRS := lldb
+OPTIONAL_DIRS := lldb, autodiff
 
 # NOTE: The tools are organized into five groups of four consisting of one
 # large and three small executables. This is done to minimize memory load
