diff --git tools/CMakeLists.txt tools/CMakeLists.txt
index b654b8c..28af50f 100644
--- tools/CMakeLists.txt
+++ tools/CMakeLists.txt
@@ -46,6 +46,7 @@ add_llvm_external_project(clang)
 add_llvm_external_project(llgo)
 add_llvm_external_project(lld)
 add_llvm_external_project(lldb)
+add_llvm_external_project(clad)
 
 # Automatically add remaining sub-directories containing a 'CMakeLists.txt'
 # file as external projects.
